File: card/middle/treesandgraphs/004/004.cpp
#include <stdlib.h>
// Populating Next Right Pointers in Each Node
// https://leetcode.com/explore/interview/card/top-interview-questions-medium/108/trees-and-graphs/789/
// https://leetcode.com/submissions/detail/172428072/


/**
 * Definition for binary tree with next pointer.
 */
struct TreeLinkNode {
    int val;
    struct TreeLinkNode *left, *right, *next;
};

void connect(struct TreeLinkNode *root) {
    
    if (root == NULL) {
        return;
    }

    struct TreeLinkNode *head = root;
    
    while(head){
        
        if (head->left) {
            head->left->next = head->right;
        }
        
        if (head->right && head->next) {
            head->right->next = head->next->left;
        }
        head = head->next;
    }
    
    connect(root->left);
}
int main(int argc, char const *argv[])
{
    return 0;
}

File: card/hard/treesandgraphs/004.go
package treesandgraphs

// Binary Tree Maximum Path Sum
// https://leetcode.com/explore/interview/card/top-interview-questions-hard/118/trees-and-graphs/845/

func max3(a, b, c int) int {
	if a > b {
		if a > c {
			return a
		}
	} else {
		if b > c {
			return b
		}
	}
	return c
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxVal := root.Val

	var loop func(*TreeNode) int
	loop = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		left := loop(root.Left)
		right := loop(root.Right)
		if left < 0 {
			left = 0
		}
		if right < 0 {
			right = 0
		}
		if left+root.Val+right > maxVal {
			maxVal = left + root.Val + right
		}
		return root.Val + max3(left, right, 0)

	}
	loop(root)

	return maxVal
}
